File: AalsiTrader/Models/BrokerPortfolio.swift
import Foundation

struct BrokerPortfolio: Codable {
    let positions: [BrokerPosition]?
    let holdings: [BrokerHolding]?
    let funds: BrokerFunds?
    let broker: String?
    let needsBrokerSetup: Bool?
    let error: String?
}

struct BrokerPosition: Codable, Identifiable {
    var id: String { tradingsymbol ?? tradingSymbol ?? UUID().uuidString }
    let tradingsymbol: String?
    let tradingSymbol: String?
    let exchange: String?
    let exchangeSegment: String?
    let quantity: Int?
    let netQty: Int?
    let average_price: Double?
    let buyAvg: Double?
    let last_price: Double?
    let lastPrice: Double?
    let pnl: Double?
    let dayPnl: Double?
    let product: String?
    let productType: String?

    var symbol: String? { tradingsymbol ?? tradingSymbol }
    var qty: Int { quantity ?? netQty ?? 0 }
    var avgPrice: Double { average_price ?? buyAvg ?? 0 }
    var currentPrice: Double { last_price ?? lastPrice ?? 0 }
    var profit: Double { pnl ?? dayPnl ?? 0 }
}

struct BrokerHolding: Codable, Identifiable {
    var id: String { tradingsymbol ?? tradingSymbol ?? UUID().uuidString }
    let tradingsymbol: String?
    let tradingSymbol: String?
    let quantity: Int?
    let average_price: Double?
    let last_price: Double?
    let pnl: Double?

    var symbol: String? { tradingsymbol ?? tradingSymbol }
    var avgPrice: Double { average_price ?? 0 }
    var currentPrice: Double { last_price ?? 0 }
    var profit: Double { pnl ?? 0 }
    var qty: Int { quantity ?? 0 }
}

struct BrokerFunds: Codable {
    let availableBalance: Double?
    let usedMargin: Double?
    let totalBalance: Double?
    let dayPnl: Double?
}

File: AalsiTraderAndroid/app/src/main/java/com/aalsitrader/android/model/BrokerPortfolio.kt
package com.aalsitrader.android.model

import kotlinx.serialization.Serializable

@Serializable
data class BrokerPortfolio(
    val positions: List<BrokerPosition>? = null,
    val holdings: List<BrokerHolding>? = null,
    val funds: BrokerFunds? = null,
    val broker: String? = null,
    val needsBrokerSetup: Boolean? = null,
    val error: String? = null,
)

@Serializable
data class BrokerPosition(
    val tradingsymbol: String? = null,
    val tradingSymbol: String? = null,
    val exchange: String? = null,
    val exchangeSegment: String? = null,
    val quantity: Int? = null,
    val netQty: Int? = null,
    val average_price: Double? = null,
    val buyAvg: Double? = null,
    val last_price: Double? = null,
    val lastPrice: Double? = null,
    val pnl: Double? = null,
    val dayPnl: Double? = null,
    val product: String? = null,
    val productType: String? = null,
) {
    val displaySymbol: String get() = tradingsymbol ?: tradingSymbol ?: ""
    val displayQty: Int get() = quantity ?: netQty ?: 0
    val displayAvgPrice: Double get() = average_price ?: buyAvg ?: 0.0
    val displayLastPrice: Double get() = last_price ?: lastPrice ?: 0.0
    val displayPnl: Double get() = pnl ?: dayPnl ?: 0.0
}

@Serializable
data class BrokerHolding(
    val tradingsymbol: String? = null,
    val tradingSymbol: String? = null,
    val quantity: Int? = null,
    val average_price: Double? = null,
    val last_price: Double? = null,
    val pnl: Double? = null,
) {
    val displaySymbol: String get() = tradingsymbol ?: tradingSymbol ?: ""
}

@Serializable
data class BrokerFunds(
    val availableBalance: Double? = null,
    val usedMargin: Double? = null,
    val totalBalance: Double? = null,
    val dayPnl: Double? = null,
)
